### PR TITLE
Improve Test coverage

### PR DESCRIPTION
**Pull request by yu-sun-77**
*Fixes #, if available:*

*Description of changes:*
Improve Test coverage from 19% to 40%

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsFixedShardsMetricsCollectorTests.java [hunk 1]
```diff
@@ -15,64 +15,93 @@
...
+        super.tearDown();
+    }
 
+    @Test
+    public void testNodeStatsMetrics() {
         try {
-            nodeStatsFixedShardsMetricsCollector.saveMetricValues("89123.23", startTimeInMills, "NodesStatsIndex", "55", "123");
-            assertTrue("Negative scenario test: Should have been a RuntimeException", true);
-        } catch (RuntimeException ex) {
-            //- expecting exception...only 3 values passed; 2 expected
```
Comment (ktkrg) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsFixedShardsMetricsCollectorTests.java:68]: Can we move this negative scenario test as well?

**Review by ktkrg — COMMENTED** [review]
Just one comment around a missing negative test scenario, looks good otherwise!

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeDetailsCollectorTests.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    DiscoveryNode testNode = new DiscoveryNode(NODE_ID, ESTestCase.buildNewFakeTransportAddress(), Collections
+        .emptyMap(),
+        DiscoveryNodeRole.BUILT_IN_ROLES, Version.CURRENT);
+
+    threadPool = new TestThreadPool("test");
+    ClusterService clusterService = ClusterServiceUtils.createClusterService(threadPool, testNode);
+    ESResources.INSTANCE.setClusterService(clusterService);
+
+    MetricsConfiguration.CONFIG_MAP.put(NodeDetailsCollector.class, MetricsConfiguration.cdefault);
+    configOverrides = Mockito.mock(ConfigOverridesWrapper.class);
```
[the remark this answers is not in the document]
Comment (yu-sun-77) [reply src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeDetailsCollectorTests.java:45]: Thanks for the comments.
In order to use `@Mock` annotation we have to `@RunWith(MockitoJUnitRunner.class)`, However as you may find these test class extends from `ESTestCase` which is running with a different `RandomizedRunner.class` and that's why I end it up using `Mockito.mock` method to make it work.

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsFixedShardsMetricsCollectorTests.java [hunk 3]
```diff
@@ -15,64 +15,93 @@
...
+        ESResources.INSTANCE.setIndicesService(indicesService);
 
-        try {
-            nodeStatsFixedShardsMetricsCollector.saveMetricValues("89123.23", startTimeInMills, "NodesStatsIndex");
-            assertTrue("Negative scenario test: Should have been a RuntimeException", true);
-        } catch (RuntimeException ex) {
-            //- expecting exception...only 1 values passed; 2 expected
-        }
+        MetricsConfiguration.CONFIG_MAP.put(NodeStatsAllShardsMetricsCollector.class, MetricsConfiguration.cdefault);
+        controller = Mockito.mock(PerformanceAnalyzerController.class);
```
[the remark this answers is not in the document]
Comment (yu-sun-77) [reply src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsFixedShardsMetricsCollectorTests.java:44]: Same reason. this test class extends from ESSingleNodeTestCase which is running with a different RandomizedRunner.class

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeDetailsCollectorTests.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yu-sun-77) [reply src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeDetailsCollectorTests.java]: Actually I realized adding `initMocks(this)` is alternative way of using `@RunWith(MockitoJUnitRunner.class)`. I've made changes to use `@Mock` annotation as suggested and it works fine now.

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsFixedShardsMetricsCollectorTests.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (yu-sun-77) [reply src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsFixedShardsMetricsCollectorTests.java]: changing to `@Mock` annotation as suggested. thanks.

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeDetailsCollectorTests.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    MetricsConfiguration.CONFIG_MAP.put(NodeDetailsCollector.class, MetricsConfiguration.cdefault);
+    collector = new NodeDetailsCollector(configOverrides);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    threadPool.shutdownNow();
+    super.tearDown();
+  }
+
```
Comment (aditjind) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeDetailsCollectorTests.java:59]: Should we add saveMetricsValues test here as well?

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsFixedShardsMetricsCollectorTests.java [hunk 5]
```diff
@@ -15,64 +15,120 @@
...
+            collector.collectMetrics(startTimeInMills);
+        } catch (Exception exception) {
+            assertTrue("There shouldn't be any exception in the code; Please check the reflection code for any changes", true);
+        }
+    }
+
+    @Test
+    public void testCollectMetrics() {
+        createIndex(TEST_INDEX);
+        Mockito.when(controller.getNodeStatsShardsPerCollection()).thenReturn(1);
```
Comment (aditjind) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsFixedShardsMetricsCollectorTests.java:99]: Should we add a test to check if all the metrics for all the shards are populated in subsequent runs of this collector?

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/CacheConfigMetricsCollectorTests.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    MetricsConfiguration.CONFIG_MAP.put(CacheConfigMetricsCollector.class, MetricsConfiguration.cdefault);
+    collector = new CacheConfigMetricsCollector();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    super.tearDown();
+  }
+
```
Comment (aditjind) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/CacheConfigMetricsCollectorTests.java:37]: Should we add saveMetricsValues test here as well?

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/CacheConfigMetricsCollectorTests.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    createIndex(TEST_INDEX);
+    collector.collectMetrics(startTimeInMills);
+
+    List<CacheMaxSizeStatus> metrics = readMetrics();
+    assertEquals(2, metrics.size());
+    CacheMaxSizeStatus filedDataCache = metrics.get(0);
+    CacheMaxSizeStatus shardRequestCache = metrics.get(1);
+    assertEquals(CacheType.FIELD_DATA_CACHE.toString(), filedDataCache.getCacheType());
+    assertEquals(CacheType.SHARD_REQUEST_CACHE.toString(), shardRequestCache.getCacheType());
```
Comment (aditjind) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/CacheConfigMetricsCollectorTests.java:50]: Should we assert values for Cache Sizes as well?

**Review by aditjind — APPROVED** [review]
I have minor non blocking comments. Approving this version, since those can be picked up in subsequent versions as well.

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/CacheConfigMetricsCollectorTests.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
+package com.amazon.opendistro.elasticsearch.performanceanalyzer.collectors;
```
Comment (ktkrg) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/CacheConfigMetricsCollectorTests.java:1]: nit: Missing license header.

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeDetailsCollectorTests.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
+package com.amazon.opendistro.elasticsearch.performanceanalyzer.collectors;
```
Comment (ktkrg) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeDetailsCollectorTests.java:1]: nit: missing license header

**Review by ktkrg — APPROVED** [review]
Just noticed that some files are missing copyright headers. Please create an issue to add them before merging this PR.

**Comment by yu-sun-77** [comment]
created https://github.com/opendistro-for-elasticsearch/performance-analyzer/issues/253 to track missing license headers for new files.

**Pull request merged**